enhancements/update/eus-upgrades-mvp.md

```diff
@@ -0,0 +1,296 @@
+authors:
+  - "@sdodson"
+reviewers:
+  - @darkmuggle
+  - @soltysh
+  - @sttts
+  - @deads2k
+  - @ecordell
+approvers:
+  - "@eparis"
+  - "@derekwaynecarr"
+  - "@crawford"
+  - "@pweil-"
+creation-date: 2020-01-12
+last-updated: 2020-01-12
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# EUS Upgrades MVP
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement outlines a set of cross platform improvements meant to ensure
+the safety of multiple back-to-back minor version upgrades associated with EUS
+to EUS upgrades. Additionally, we outline improvements applicable to all
+upgrades meant to reduce the duration of upgrades and workload disruption.
+
+All of the work detailed herein is pre-requisite for any future attempts to skip
+upgrade steps such as reboots between specific minor releases.
+
+## Motivation
+
+The introduction of EUS creates a subset of clusters which we expect will run
+4.6 for a year or more then upgrade rapidly, though serially, from 4.6 to 4.10.
+This rapid upgrade introduces the risk that those clusters may upgrade faster
+than is safe due constraints imposed by OpenShift, the upstream components of
+OpenShift, or deployed workloads.
+
+This also creates a scenario where admins wish to reduce both the duration and
+the disruption to workload associated with the upgrade.
+
+### Goals
+
+- We will inform admins of incompatibilities between their current usage and the
+final upgrade target
+- We will provide guidance on expected upgrade duration scaled by node count exclusive
+of the variability associated with workload rescheduling (ie: everything except
+Worker MCP rollout)
+- We will inhibit upgrades to the next minor version whenever necessary to preserve
+cluster stability and supportability, ie:
+  - Respecting version skew policies between OS managed components (kubelet, crio, RHCOS)
+  and operator managed components (Kubernetes API, OpenShift API, etc)
+  - Respecting version skew policies of OLM managed Operators
+- We will ensure that CVO managed operators upgrade as quickly as is safe to do so
+- We will reduce workload disruption associated with pod rescheduling during rolling
+reboots
+
+
+### Non-Goals
+
+- This enhancement does not attempt to remove *any* steps along the serial
+  upgrade path from 4.6 to 4.7 to 4.8 to 4.9 to 4.10 *including* reboots. A
+  follow-up enhancement will address those concerns.
+
+## Proposal
+
+### User Stories
+
+#### As an admin preparing for 4.6 to 4.10 EUS upgrade I'm alerted to API removals
+
+We have pre-existing mechanisms to alert admins to their use of APIs which are
```

Comment:
I'm assuming this is the notification and alert section.

For UX related teams: I think we need to clarify the experience on if this would be incorporated to part of the upgrade visualization (in addition to alerting) and if so how would that API work?

Reply:
It may be useful if the CVO scraped up alerts of this nature and raised them and then the console could only display alerts firing from CVO on the upgrades panel.

Reply:
Do we need to consider using upgradeable=false if we detect usage of an API that is removed in next version?

iow.:

4.6
4.7
4.8: API server go "upgradeable=false" if there are cronjobs in the cluster
4.9: CronJobs are gone, but cluster has still 1000 cron jobs created
4.10:

Reply:
Info alerts are planned for 4.7, showing up in the console notification side-bar. We will have to backport to 4.6.

Reply:
@sdodson interesting idea from @sferich888 today: if the APIs are deprecated in what we generate our API docs from can we use that info to also generate big, scary release notes?

Another suggestion for collaboration: @jwmatthews and the container migration teams may have thoughts on how we can help admins work through deprecations and other ways to make sure they're front and center during our upgrade process.

Reply:
@pweil- FYI, we have some beginning validation work in context of Migrations from MTC to warn a user when they will face a problem that a GVK they relied on in the source cluster isn't present in the target cluster (https://github.com/konveyor/mig-controller/blob/master/pkg/gvk/gvk.go#L66)  This is a problem space we are interested in, we expect to see it with future migrations, open to collaborating.    (cc @djwhatle @eriknelson @shawn-hurley @jortel )

Reply:
k8s also has https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1693-warnings which kubectl & oc consume informing users that they are using deprecating APIs. The question is if/how we could also consume that elsewhere.

Reply:
@pweil- @sferich888  Yeah I definitely agree that we'll want to have a release notes if not an upgrade section of docs that's unique to 4.6 to 4.10 upgrades. I'll add an epic place holder to ensure that as part of 4.10 dev cycle we pull this together.

I'll find a slot next week to sync up with @sanchezl @deads2k and you, @jwmatthews.